Implement Multi-Operation Endpoint for Subbasin Modeling

rajadain opened the pull request:
## Overview

Implements the MultiOperation outlined in #80 and #82. This accepts a JSON request specifying multipls shapes and operations, performs all of them, and returns a single JSON result.

Connects #80 

### Demo

Some initial timing:

```fish
$ for i in 1 2 3 4 5; time -p http --timeout=90 :8090/multi < examples/MultiOperationRequest.json 2>&1 > /dev/null | grep real | awk '{print $2}'; end

33.89
21.79
19.57
21.12
18.59
```

### Notes

I expect this PR to go through multiple revisions as we tweak performance.

The sample output was generated by:

```bash
$ http --timeout 90 :8090/multi < examples/MultiOperationRequest.json | jq . > examples/MultiOperationResponse.json
```

## Testing Instructions

* Checkout this branch and run `server`
* Do:

    ```bash
    http --timeout 90 :8090/multi < examples/MultiOperationRequest.json
    ```

  and see the result.

rajadain committed:
Naive implementation of multi-operation endpoint

rajadain committed:
Remove Mocks

rajadain committed:
Add sample response

rajadain commented on api/src/main/resources/application.conf:
```diff
@@ -5,6 +5,9 @@ geoprocessing {
 
 akka.http {
+  server {
+    request-timeout = infinite
```
We probably should not be doing this. The new end point does take on average 22s, the highest I've recorded is 35s, which is quite a bit more than the 20s default limit of Akka Http. I'm hoping we can do some tweaks to lower the time of the request, or at least add a more educated cap than `infinite`.

rajadain commented on api/src/main/scala/Geoprocessing.scala:
```diff
@@ -13,917 +13,45 @@ import geotrellis.vector._
+  @throws(classOf[MissingTargetRasterException])
+  def getMultiOperations(input: MultiInput): Map[String, Map[String, Map[String, Double]]] = {
+    val hucs = input.shapes.map(huc => huc.id -> normalizeHuc(huc))
+    val rasters = collectRastersForInput(input, hucs.map(_._2))
```
This crops all distinct layers mentioned in the request to the union of all the given shapes. Can we use a `Future` here or something?

rajadain commented on api/src/main/scala/Geoprocessing.scala:
```diff
@@ -13,917 +13,45 @@ import geotrellis.vector._
+      val layers = op.rasters.map(r => rasters(r))
+      val result = op.name match {
+        case "RasterGroupedCount" =>
+          rasterGroupedCount(layers, shape, opts).mapValues(_.toDouble)
```
This and `rasterLinesJoin` below return `Map[String, Int]` but to match the type signature of this method I have to convert them to `Double`. Is there a type signature we can use that doesn't require this?

rajadain reviewed (commented):
Added some specific questions about the implementation.

rajadain committed:
GeoTrellis Optimizations

echeipesh committed:
GeoTrellis Cleanup

rajadain committed:
Civic Apps Cleanup

rajadain commented:
Here's the comparison chart for the latest cleanup:

![image](https://user-images.githubusercontent.com/1430060/37478651-3366590e-2851-11e8-9fbf-d62bfdaa309d.png)

with the final results being the average. While "Civic Apps Cleanup" does take slightly longer, I'm not sure what the cause of it in the code, since the type aliases should be handled at compile time. The only runtime code change I made was changing this `map` to `mapValues`:

```scala
val result1: Map[String, Future[Map[String,Map[String,Double]]]] =
  result0.map { case (k, v) => k -> sequenceMap(v) }
```

to

```scala
val result1: Map[HucID, Future[Map[OperationID, Map[String, Double]]]] =
  result0.mapValues(sequenceMap)
```

and even though there is [some discussion on `map` vs `mapValues`](http://blog.bruchez.name/2013/02/mapmap-vs-mapmapvalues.html) it doesn't seem to allege performance differences. I'm chalking it up to noise.

This is ready for one final review.

moradology committed:
Possible cleanup

rajadain committed:
Merge pull request #86 from moradology/npz/subbasin-cleanup

MultiInput GT cleanup

arottersman reviewed (approved):
+1, tested...real nice job thinking through and corralling this work, @rajadain. Google charts are also a great help. And, thanks GT for the great pairing!

After three "__ Clean Ups" code is looking pretty squeaky clean. Very excited to have this wired into mapshed.

rajadain committed:
Remove unused utility functions

These were introduced at various points in this feature,
but ended up not being used in the final version.

rajadain commented:
Thanks for all the input! This turned out great. Added a final commit to remove the extraneous utility functions we ended up not using (they're safe in the git history should we ever need to look back to them). Merging now.
